python_modules/dagster/dagster/core/execution/context/system.py

```diff
@@ -720,27 +721,39 @@ def op_config(self) -> Any:
         return solid_config.config if solid_config else None
 
     def has_asset_partitions_for_input(self, input_name: str) -> bool:
-        op_config = self.op_config
-
-        if is_iterable(op_config) and "assets" in op_config:
-            all_input_asset_partitions = op_config["assets"].get("input_partitions")
-            if all_input_asset_partitions is not None:
-                this_input_asset_partitions = all_input_asset_partitions.get(input_name)
-                if this_input_asset_partitions is not None:
-                    return True
+        asset_layer = self.pipeline_def.asset_layer
+        assets_def = asset_layer.assets_def_for_asset(
+            next(iter(asset_layer.asset_keys_for_node(self.solid_handle)))
+        )
+        upstream_asset_key = asset_layer.asset_key_for_input(self.solid_handle, input_name)
 
-        return False
+        return (
+            upstream_asset_key is not None
+            and assets_def.partitions_def is not None
+            and asset_layer.partitions_def_for_asset(upstream_asset_key) is not None
+        )
 
     def asset_partition_key_range_for_input(self, input_name: str) -> PartitionKeyRange:
-        op_config = self.op_config
-        if is_iterable(op_config) and "assets" in op_config:
-            all_input_asset_partitions = op_config["assets"].get("input_partitions")
-            if all_input_asset_partitions is not None:
-                this_input_asset_partitions = all_input_asset_partitions.get(input_name)
-                if this_input_asset_partitions is not None:
-                    return PartitionKeyRange(
-                        this_input_asset_partitions["start"], this_input_asset_partitions["end"]
-                    )
+        from dagster.core.asset_defs.asset_partitions import (
+            get_upstream_partitions_for_partition_range,
+        )
+
+        asset_layer = self.pipeline_def.asset_layer
+        assets_def = asset_layer.assets_def_for_asset(
+            next(iter(asset_layer.asset_keys_for_node(self.solid_handle)))
```

Comment:
we'll definitely want to do a big ol' refactor of the AssetLayer class to make stuff like this less annoying